[FAB-17605] Handling errors returned by InitFactories()

Eggsyz opened the pull request:
<!--- DELETE MARKDOWN COMMENTS BEFORE SUBMITTING PULL REQUEST. -->

<!--- Provide a descriptive summary of your changes in the Title above. -->

#### Type of change

<!--- What type of change? Pick one option and delete the others. -->

- Bug fix

#### Description

<!--- Describe your changes in detail, including motivation. -->
A project using func InitFactories, it output errors when build it. But many places do not handle errors.

Signed-off-by: jilg <liguo_ji@outlook.com>

#### Additional details

<!--- Additional implementation details or comments to reviewers. -->
<!--- Summarize how the pull request was tested (if not obvious from commit). -->

#### Related issues

<!--- Include a link to any associated issues, e.g. Jira issue or approved rfc. -->
[17605](https://jira.hyperledger.org/browse/FAB-17605?filter=-1)
<!---
#### Release Note
If change impacts current users, uncomment Release Note heading and provide
release note text.
Also, copy release note text into the release specific /release_notes file.
-->

<!--
Checklist (DELETE AFTER READING):

- `Signed-off-by` added to commits (required for DCO check to pass)
- Tests have been added/updated (required for bug fixes and features)
- Unit and/or integration tests pass locally
- Run linters and checks locally using 'make checks'
- If change requires documentation updates, make updates in pull request,
  or open a separate issue and provide link
- Squash commits into a single commit, unless a stack of commits is
  intentional to assist reviewers or to preserve review comments.
- For additional contribution guidelines see the project's CONTRIBUTING.md file
-->

denyeart commented on gossip/gossip/gossip_test.go:
```diff
@@ -42,7 +42,10 @@ func TestMain(m *testing.M) {
-	factory.InitFactories(nil)
+	err := factory.InitFactories(nil)
+	if err != nil {
+		panic(fmt.Errorf("Could not initialize BCCSP Factories [%s]", err))
```
assert.NoError() here as well.

Eggsyz replied:
this is testing.M, not testing.T

denyeart commented on internal/configtxgen/encoder/encoder_suite_test.go:
```diff
@@ -15,8 +16,10 @@ import (
-
+	err := factory.InitFactories(nil)
+	if err != nil {
+		panic(fmt.Errorf("Could not initialize BCCSP Factories [%s]", err))
```
assert.NoError() here as well.

MHBauer reviewed (dismissed):
@denyeart how to get a `testing.T` in the `init()` functions?

denyeart reviewed (commented):
Ah, you're right, removed the comment from the init functions.

Eggsyz committed:
[FAB-17605] Handling errors returned by InitFactories()

A project using func InitFactories, it output errors when build it. But many places do not handle errors.

Signed-off-by: jilg <liguo_ji@outlook.com>

denyeart commented:
I'll defer to @sykesm on whether the panics are the most appropriate way to handle the errors.

sykesm commented:
The *vast* majority of calls to InitFactory simply pass in a nil configuration. If that function does not work, it's pretty much always fatal as there's little on the default config path that can go wrong; it's wired to work. The error return is only for cases where an explicit configuration is provided.

The places where we pass in an explicit configuration, the call can fail. It looks like there are only two places where that is the case and only one of those is production code:

- `msp/configbuilder.go`
- `core/testutil/config.go`

The orderer and peer implementations initialize bccsp factories by side-effect through direct or indirect calls to `msp.GetLocalMspConfig` so that's the only place where an error really can happen.

In the peer case, the path is through `InitCrypto` and, if that call fails, the process is terminated with an `os.Exit(1)`. In the orderer case, `GetLocalMspConfig` is called directly and, if that call fails, the process is terminated with a panic.

So... IMO, given how poorly structured bccsp is, if `InitFactories` fails, it should simply panic. The function shouldn't bother returning an error. While I generally prefer error returns to panic, crypto is so fundamentally core to Fabric that failures to set it up due to problems in the configuration or environment are always terminal. We should codify that.
